common/src/main/java/org/apache/uniffle/common/merger/Merger.java

```
@@ -0,0 +1,208 @@
+        }
+      }
+      initialize(segmentsToMerge.size());
+      clear();
+      for (Segment segment : segmentsToMerge) {
+        put(segment);
+      }
+    }
+
+    @Override
+    public Object getCurrentKey() {
+      return currentKey;
+    }
+
+    @Override
+    public Object getCurrentValue() {
+      return currentValue;
+    }
+
+    @Override
+    public boolean next() throws IOException {
+      if (size() == 0) {
+        resetKeyValue();
+        return false;
+      }
+
+      if (minSegment != null) {
+        adjustPriorityQueue(minSegment);
+        if (size() == 0) {
+          minSegment = null;
+          resetKeyValue();
+          return false;
+        }
+      }
+      minSegment = top();
+      currentKey = minSegment.getCurrentKey();
+      currentValue = minSegment.getCurrentValue();
+      return true;
+    }
+
+    private void resetKeyValue() {
+      currentKey = null;
+      currentValue = null;
+    }
+
+    private void adjustPriorityQueue(Segment segment) throws IOException {
+      if (segment.next()) {
+        adjustTop();
+      } else {
+        pop();
+        segment.close();
+        if (popSegmentHook != null) {
+          Segment newSegment = popSegmentHook.apply((int) segment.getId());
+          if (newSegment != null) {
+            if (newSegment.next()) {
+              put(newSegment);
+            } else {
+              newSegment.close();
+            }
+          }
+        }
+      }
+    }
+
+    void merge(OutputStream output) throws IOException {
+      RecordsWriter<K, V> writer =
+          new RecordsWriter<K, V>(rssConf, output, keyClass, valueClass, raw);
+      boolean recorded = true;
+      while (this.next()) {
+        writer.append(this.getCurrentKey(), this.getCurrentValue());
+        if (output instanceof Recordable) {
+          recorded =
+              ((Recordable) output)
+                  .record(writer.getTotalBytesWritten(), () -> writer.flush(), false);
+        }
```

We should throw an exception?